kernel/src/hil/time.rs

```diff
@@ -2,20 +2,31 @@
 
 use crate::ReturnCode;
 
-pub trait Time<W = u32> {
+pub trait Time {
+    type Ticks: Ticks;
     type Frequency: Frequency;
 
     /// Returns the current time in hardware clock units.
-    fn now(&self) -> W;
+    fn now(&self) -> Self::Ticks;
 
-    /// Returns the wrap-around value of the clock.
-    ///
-    /// The maximum value of the clock, at which `now` will wrap around. I.e., this should return
-    /// `core::u32::MAX` on a 32-bit-clock, or `(1 << 24) - 1` for a 24-bit clock.
-    fn max_tics(&self) -> W;
+    fn ticks_from_seconds(s: u32) -> Self::Ticks {
```

Comment:
I'm a little wary of these conversion routines, particularly because ticks_from_ms(ticks_to_ms(x)) won't necessarily compute the identity, and because there are floating point operations. 

It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.

Reply:
> I'm a little wary of these conversion routines, particularly because ticks_from_ms(ticks_to_ms(x)) won't necessarily compute the identity, and because there are floating point operations.

1. `ticks_from_ms(ticks_to_ms(x))` will definitely pretty much never compute the identity, whatever the formula is (using floating point or integer arithmetic), unless the frequency is a "nice" number (like 1, or powers of two for "reasonable" values of x that don't trigger an overflow), or unless `ticks_to_ms(x)`'s result stores the original `x`. But is it unexpected? If one wants to keep the precision, the operations must be done in ticks.
Actually, the only use of `ticks_to_ms` I found was in `capsules/src/ieee802154/xmac.rs`, and it's questionable whether this conversion is needed there, but I don't have knowledge of this protocol & driver to assess that.
The other uses of `ticks_from_ms` all sound reasonable to me (most seem like heuristics, or "average" values from a data sheet). It's unclear though whether the value should be rounded up instead of down, so that the waiting time is indeed at least N ms, instead of N-1 < x <= N.
1. There was a lot of code duplication so I took the approach of consistently using floating point arithmetic to avoid overflows on multiplication. Given that at least one driver was using floating point arithmetic I assumed that it was fine, but we could also use integer arithmetic instead, or provide two variants of the function.
The caveat of integer arithmetic is that "ms * frequency" can overflow before dividing by 10^3. On the other hand, one can divide by 10^3 before multiplying by the frequency, but then the precision is rounded to full seconds. A more accurate formula is `(ms / 1000) * f + ((ms % 1000) * f) / 1000` - assuming that `1000 * f` doesn't overflow (if you convert from microseconds instead of milliseconds it will overflow).
Another alternative is to use 64-bit arithmetic. It can be evaluated even on 32-bit CPUs (at the expense of a few more instructions), and given that the divisor is a compile-time constant, the compiler should be able to turn the division into a combination of multiplication(s), shift(s) and other simple arithmetic operations (example, on x86_64: https://godbolt.org/z/7nvfWw). Also, with a compile-time constant number of milliseconds (which is the most common case), the compiler can just evaluate that and bake a constant in the binary. So 64-bit arithmetic would be the best alternative to floating point IMO

What do you think?

> It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.

That would be possible, although out of the scope of this pull-request. The core goal of this pull-request is fixing 24-bit ticks, so that the alarms based on nRF52's RTC work as intended, and it's already a large enough change to fix an orthogonal issue.

Reply:
>    Actually, the only use of `ticks_to_ms` I found was in `capsules/src/ieee802154/xmac.rs`, and it's questionable whether this conversion is needed there, but I don't have knowledge of this protocol & driver to assess that.

Yeah, this is exactly the kind of use I was worried about. The protocol is trying to figure out when to turn the radio on. The more accurately it can do this, the less energy it will use. What the code does is take the ticks remaining, turn them into ms, then convert back from ms to ticks. I think it's pretty simple to fix this (change `set_timer_ms` to `set_timer_ticks` and just use ticks), and in that case there's no need for `ticks_to_ms`. 

>    The other uses of `ticks_from_ms` all sound reasonable to me (most seem like heuristics, or "average" values from a data sheet). It's unclear though whether the value should be rounded up instead of down, so that the waiting time is indeed at least N ms, instead of N-1 < x <= N.

Round up. Or just round.

> 2. There was a lot of code duplication so I took the approach of consistently using floating point arithmetic to avoid overflows on multiplication. Given that at least one driver was using floating point arithmetic I assumed that it was fine, but we could also use integer arithmetic instead, or provide two variants of the function.
>    The caveat of integer arithmetic is that "ms * frequency" can overflow before dividing by 10^3. On the other hand, one can divide by 10^3 before multiplying by the frequency, but then the precision is rounded to full seconds. A more accurate formula is `(ms / 1000) * f + ((ms % 1000) * f) / 1000` - assuming that `1000 * f` doesn't overflow (if you convert from microseconds instead of milliseconds it will overflow).
>    Another alternative is to use 64-bit arithmetic. It can be evaluated even on 32-bit CPUs (at the expense of a few more instructions), and given that the divisor is a compile-time constant, the compiler should be able to turn the division into a combination of multiplication(s), shift(s) and other simple arithmetic operations (example, on x86_64: https://godbolt.org/z/7nvfWw). Also, with a compile-time constant number of milliseconds (which is the most common case), the compiler can just evaluate that and bake a constant in the binary. So 64-bit arithmetic would be the best alternative to floating point IMO
> 
> What do you think?

64 bit is the way to go. I think we encountered something like this before in timer code (I'm struggling to find where) and the conclusion was to use 64-bit arithmetic rather than do tricks with 32-bit arithmetic.

> > It also can lead to runtime failures, because not every tick frequency can support every unit. For example, if the Alarm is based off a 32kHz clock and code requests a 10us Alarm. I realize this problem existed with the previous trait as well. Is there any way to parameterize the Alarm trait by the tick unit, such that, for example, the UDP test has a ms alarm but I2C has a us alarm? A ms alarm can just be a scaling of the us one.
> 
> That would be possible, although out of the scope of this pull-request. The core goal of this pull-request is fixing 24-bit ticks, so that the alarms based on nRF52's RTC work as intended, and it's already a large enough change to fix an orthogonal issue.

I agree.

Reply:
Migrated to `u64` arithmetic, and removed `ticks_to_ms`. Please confirm that the `ieee802154/xmac` capsule is still correct.

Reply:
hrm, could we use `core::time::Duration` instead of `ticks_from_xx` instead?

Reply:
`u64` is definitely better than floating point (the code bloat alone from software floating point would be a huge problem a big problem on non-M4F platforms).

I agree with @niklasad1 that if there isn't a reason _not_ to use `Duration` it probably makes sense to unify all of these with `ticks_from_duration`.

Finally, this is also a candidate for being a helper as opposed to a method.